crypto/x509/x509_vfy.c

```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
+         */
+        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
+            && !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
+                 && (x->ex_flags & EXFLAG_SI) != 0)) {
+            /* Check Basic Constraints according to RFC 5280 section 4.2.1.9 */
+            if (x->ex_pathlen != -1) {
+                if ((x->ex_flags & EXFLAG_CA) == 0)
+                    ctx->error = X509_V_ERR_PATHLEN_INVALID_FOR_NON_CA;
+                if ((x->ex_kusage & KU_KEY_CERT_SIGN) == 0)
+                    ctx->error = X509_V_ERR_PATHLEN_WITHOUT_KU_KEY_CERT_SIGN;
+            }
+            if ((x->ex_flags & EXFLAG_CA) != 0
+                    && (x->ex_flags & EXFLAG_BCONS) != 0
+                    && (x->ex_flags & EXFLAG_BCONS_CRITICAL) == 0)
+                ctx->error = X509_V_ERR_CA_BCONS_NOT_CRITICAL;
+            /* Check Key Usage according to RFC 5280 section 4.2.1.3 */
+            if ((x->ex_flags & EXFLAG_CA) != 0) {
+                if ((x->ex_flags & EXFLAG_KUSAGE) == 0)
+                    ctx->error = X509_V_ERR_CA_CERT_MISSING_KEY_USAGE;
+            } else {
+                if ((x->ex_kusage & KU_KEY_CERT_SIGN) != 0)
+                    ctx->error = X509_V_ERR_KU_KEY_CERT_SIGN_INVALID_FOR_NON_CA;
+            }
+            /* Check issuer is non-empty acc. to RFC 5280 section 4.1.2.4 */
+            if (X509_NAME_entry_count(X509_get_issuer_name(x)) == 0)
+                ctx->error = X509_V_ERR_ISSUER_NAME_EMPTY;
+            /* Check subject is non-empty acc. to RFC 5280 section 4.1.2.6 */
+            if (((x->ex_flags & EXFLAG_CA) != 0
+                 || (x->ex_kusage & KU_CRL_SIGN) != 0
+                 || x->altname == NULL
+                 ) && X509_NAME_entry_count(X509_get_subject_name(x)) == 0)
+                ctx->error = X509_V_ERR_SUBJECT_NAME_EMPTY;
+            if (X509_NAME_entry_count(X509_get_subject_name(x)) == 0
+                    && x->altname != NULL
+                    && (x->ex_flags & EXFLAG_SAN_CRITICAL) == 0)
+                ctx->error = X509_V_ERR_EMPTY_SUBJECT_SAN_NOT_CRITICAL;
+            /* Check SAN is non-empty according to RFC 5280 section 4.2.1.6 */
+            if (x->altname != NULL && sk_GENERAL_NAME_num(x->altname) <= 0)
+                ctx->error = X509_V_ERR_EMPTY_SUBJECT_ALT_NAME;
+            /* TODO add more checks on SAN entries */
+            /* Check sig alg consistency acc. to RFC 5280 section 4.1.1.2 */
+            if (X509_ALGOR_cmp(&x->sig_alg, &x->cert_info.signature) != 0)
+                ctx->error = X509_V_ERR_SIGNATURE_ALGORITHM_INCONSISTENCY;
+            if (x->akid != NULL && (x->ex_flags & EXFLAG_AKID_CRITICAL) != 0)
+                ctx->error = X509_V_ERR_AUTHORITY_KEY_IDENTIFIER_CRITICAL;
+            if (x->akid != NULL && (x->ex_flags & EXFLAG_SKID_CRITICAL) != 0)
+                ctx->error = X509_V_ERR_SUBJECT_KEY_IDENTIFIER_CRITICAL;
+            if (X509_get_version(x) >= 2) { /* at least X.509v3 */
```

Comment:
I think we should just generate an error in case it's not at least a v3, I don't see a reason to only do that check for v3.

Reply:
As far as I understand, v1 or v2 certs cannot have extensions such as SKID and AKID,
so without that guard we would always report non-conformance for certs not having at least version 3.

Reply:
You're right that in 4.1.2.1 it actually allows version 1 and 2 certificates in special cases. But for a CA we require that it has a key usage extension above. The AKID is also a MUST for non-root CAs in the RFC.

Reply:
> You're right that in 4.1.2.1 it actually allows version 1 and 2 certificates in special cases.

Yes.

> But for a CA we require that it has a key usage extension above.

We check for existence of key usage only for certs with EXFLAG_CA, which can only be set if a basic key usage extension is present, which can happen only for certs with version at least 3. So here we do not need the explicit guard `X509_get_version(x) >= 2`.

> The AKID is also a MUST for non-root CAs in the RFC.

That's right, and this test is included as part of this PR (or course, only for at least v3 as discussed).

Reply:
I was under the impression that EXFLAG_CA was set in all cases X509_check_ca said it was a CA, but clearly it's only when the basic constrains section exists.

Do we have an places where we enforce that extensions are only allowed for v3 certs?

Reply:
Good point - such a check was missing; I've just added it.